src/libraries/System.Security.Cryptography.Algorithms/tests/ChaCha20Poly1305Tests.cs

```diff
@@ -0,0 +1,464 @@
+        {
+            using (var chaChaPoly = new ChaCha20Poly1305(testCase.Key))
+            {
+                byte[] ciphertext = new byte[testCase.Plaintext.Length];
+                byte[] tag = new byte[testCase.Tag.Length];
+                chaChaPoly.Encrypt(testCase.Nonce, testCase.Plaintext, ciphertext, tag, testCase.AssociatedData);
+                Assert.Equal(testCase.Ciphertext, ciphertext);
+                Assert.Equal(testCase.Tag, tag);
+
+                byte[] plaintext = new byte[testCase.Plaintext.Length];
+                chaChaPoly.Decrypt(testCase.Nonce, ciphertext, tag, plaintext, testCase.AssociatedData);
+                Assert.Equal(testCase.Plaintext, plaintext);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc8439TestCases))]
+        public static void Rfc8439TestsTamperTag(AEADTest testCase)
+        {
+            using (var chaChaPoly = new ChaCha20Poly1305(testCase.Key))
+            {
+                byte[] ciphertext = new byte[testCase.Plaintext.Length];
+                byte[] tag = new byte[testCase.Tag.Length];
+                chaChaPoly.Encrypt(testCase.Nonce, testCase.Plaintext, ciphertext, tag, testCase.AssociatedData);
+                Assert.Equal(testCase.Ciphertext, ciphertext);
+                Assert.Equal(testCase.Tag, tag);
+
+                tag[0] ^= 1;
+
+                byte[] plaintext = RandomNumberGenerator.GetBytes(testCase.Plaintext.Length);
+                Assert.Throws<CryptographicException>(
+                    () => chaChaPoly.Decrypt(testCase.Nonce, ciphertext, tag, plaintext, testCase.AssociatedData));
+                Assert.Equal(new byte[plaintext.Length], plaintext);
+            }
+        }
+
+        public static IEnumerable<object[]> GetInvalidNonceSizes()
+        {
+            yield return new object[] { 0 };
+            yield return new object[] { 8 };
+            yield return new object[] { 11 };
+            yield return new object[] { 13 };
+            yield return new object[] { 16 };
+        }
+
+        public static IEnumerable<object[]> GetInvalidTagSizes()
+        {
+            yield return new object[] { 0 };
+            yield return new object[] { 8 };
+            yield return new object[] { 12 };
+            yield return new object[] { 15 };
+            yield return new object[] { 17 };
+        }
+
+        // https://tools.ietf.org/html/rfc8439
+        private const string Rfc8439TestVectors = "RFC 8439 Test Vectors";
+
+        public static IEnumerable<object[]> GetRfc8439TestCases()
+        {
+            foreach (AEADTest test in s_rfc8439TestVectors)
+            {
+                yield return new object[] { test };
+            }
+        }
+
+        // CaseId is unique per test case
+        private static readonly AEADTest[] s_rfc8439TestVectors = new AEADTest[]
+        {
+            new AEADTest
+            {
+                Source = Rfc8439TestVectors,
+                CaseId = 1,
```

Comment:
Since at least cases 1 and 2 aren't from the same section, it'd be nice to link to the places they came from in a comment.

I think the second is https://tools.ietf.org/html/rfc8439#appendix-A.5, and the first is https://tools.ietf.org/html/rfc8439#section-2.4.2

Reply:
Sure. BTW if you have any other ideas for where I can find test vectors, I'm all ears. The RFC isn't exactly flush with them. :/

For reference, symcrypt [uses its own test vectors](https://github.com/microsoft/SymCrypt/blob/6d019cefafb3fefe3c53b0de3bba6f8c86e2d48a/lib/chacha20_poly1305.c#L219-L255).